### DISCUSSION TITLE
My own module and dependeny of netbeans module (friend issue)

### DISCUSSION THREAD

**Discussion by 6PATyCb** [apache/netbeans#4753] [Q&A] [answered]
Hi! I try to make my own nb module for support proprietary task repository. As template I am using [netbeans-github-issues-plugin](https://github.com/junichi11/netbeans-github-issues-plugin). When I try to use visual  [component](https://github.com/apache/netbeans/blob/master/ide/team.commons/src/org/netbeans/modules/bugtracking/commons/LinkButton.java) from public package of `Team Commons` module, I catch `class not found exception` when my module is running:
![image](https://user-images.githubusercontent.com/33724393/194655793-b605a0bd-fa41-42d5-b164-2d8f3b197e7f.png)
This module was added to my pom.xml:
![image](https://user-images.githubusercontent.com/33724393/194656946-6d915359-00f4-4018-acda-f6b02f5ab085.png)
On compile I catch warning:

> Module has friend dependency on org.netbeans.modules.team.commons but is not listed as a friend.

And on runtime i catch this:
![image](https://user-images.githubusercontent.com/33724393/194659224-a77a1df8-0289-4ee9-a37c-add1035c5fef.png)

I can't figure out if I only use interfaces of dependent public nb modules - everything works fine, but after I try to add class dependency - everything crashes.

In `Team Commons` i found mark as friend configuration for `netbeans-github-issues-plugin`:
![image](https://user-images.githubusercontent.com/33724393/194670629-90b6227f-9d04-46d2-b4ef-fdcde75a0c56.png)
Is this a hack to solve a `friend` exception, or is there another legal way? Maybe there is a legal config for `nbm-maven-plugin`, but I couldn't google it?

**Comment by mbien** [comment, marked as answer]
As external plugin you won't be able to access those packages unless you are registered as [friend](https://netbeans.apache.org/wiki/DevFaqImplementationDependency#_friend_dependencies) in the `team.commons` module.

You can use the [implementation dependency](https://netbeans.apache.org/wiki/DevFaqImplementationDependency#_what_is_an_implementation_dependency_and_when_should_i_use_one), but this would require to create a new build of your plugin every time there is a new NetBeans release.

**Comment by 6PATyCb** [comment]
As soon as I asked a question, I myself was able to answer it. The rubber duck effect in action :)

> You can use the [implementation dependency](https://netbeans.apache.org/wiki/DevFaqImplementationDependency#_what_is_an_implementation_dependency_and_when_should_i_use_one), but this would require to create a new build of your plugin every time there is a new NetBeans release.

The idea became clear. It's better not to create unnecessary cohesion with the implementation of the netbeans itself, but if you still decide to do this, then I'll leave it here:

1. Find the module you are going to depend on:
![image](https://user-images.githubusercontent.com/33724393/194752458-14b11acc-c11b-412c-b9b3-d77feeb560cc.png)

2. Open it with zip and find manifest file:
![image](https://user-images.githubusercontent.com/33724393/194752945-57603702-0af5-452c-a6b7-25da1043159b.png)

3. Move information from manifest to your pom.xml:
![image](https://user-images.githubusercontent.com/33724393/194753122-2587a7f2-28e4-4c89-b7b1-a2cfe1394d17.png)

4. Don't forget use right version of dependency  in pom.xml:
![image](https://user-images.githubusercontent.com/33724393/194753317-46b9db4a-98bf-4f38-82e2-64402898bdfa.png)
